internal/typeinfo/info.go

```diff
@@ -13,17 +13,131 @@ import (
 // the parser.
 var validColNameRx = regexp.MustCompile(`^([a-zA-Z_])+([a-zA-Z_0-9])*$`)
 
-// Info exposes useful information about types used in SQLair queries.
-type Info interface {
-	Typ() reflect.Type
+// ArgInfo is used to access type information about SQLair input and output
+// arguments. Methods on ArgInfo can be used to generate input and output value
+// locators.
+//
+// ArgInfo should only be accessed using it methods, not used directly as a
+// map.
+type ArgInfo map[string]arg
+
+// GenerateArgInfo takes sample instantiations of argument types and uses
+// reflection to generate an ArgInfo containing the types.
+func GenerateArgInfo(typeSamples ...any) (ArgInfo, error) {
```

Comment:
Variadic arg lists are ergonomic on an exported API, but once inside internal/private packages just make them slices. Then you don't have to keep splatting through the call stack.